application/test/models/project_test.rb

```diff
@@ -0,0 +1,231 @@
+require "test_helper"
+
+class ProjectTest < ActiveSupport::TestCase
+
+  test "initialization should works" do
+    target = Project.new(id: 'ab12345', name: 'test_project', download_dir: '/tmp/test_project')
+    assert_equal 'ab12345', target.id
+    assert_equal 'test_project', target.name
+    assert_equal '/tmp/test_project', target.download_dir
+  end
+
+  test "should be valid when all fields are populated" do
+    target = create_valid_project
+    assert target.valid?
+  end
+
+  test "validations should fail due to blank value" do
+    target = create_valid_project
+    assert target.valid?
+
+    target.id = ''
+    refute target.valid?
+    assert_includes target.errors[:id], "can't be blank"
+
+    target.name = ''
+    refute target.valid?
+    assert_includes target.errors[:name], "can't be blank"
+
+    target.download_dir = ''
+    refute target.valid?
+    assert_includes target.errors[:download_dir], "can't be blank"
+  end
+
+  test "to_hash" do
+    target = create_valid_project
+    expected_hash = {id: target.id, name: target.name, download_dir: target.download_dir}.stringify_keys
+    assert_equal expected_hash, target.to_hash
+  end
+
+  test "to_json" do
+    target = create_valid_project
+    expected_json = {id: target.id, name: target.name, download_dir: target.download_dir}.to_json
+    assert_equal expected_json, target.to_json
+  end
+
+  test "to_yaml" do
+    target = create_valid_project
+    expected_yaml = {id: target.id, name: target.name, download_dir: target.download_dir}.stringify_keys.to_yaml
+    assert_equal expected_yaml, target.to_yaml
+  end
+
+  test "save with valid attributes" do
+    in_temp_directory do |dir|
+      target = create_valid_project
+      assert target.save
+      expected_file = File.join(dir, 'projects', target.id, 'metadata.yml')
+      assert File.exist?(expected_file), "Downloadproject file was not created in the file system"
```

Comment:
should we put in the messages of this test class `DownloadProject` or `Project`?

Reply:
project best - will update